realsense_camera/src/base_nodelet.cpp

```diff
@@ -514,68 +517,56 @@ namespace realsense_camera
+  */
+  void BaseNodelet::setFrameCallbacks()
   {
-    // Enable streams.
-    if (enable_[RS_STREAM_COLOR] == true)
+    depth_frame_handler_ = [&](rs::frame  frame)
     {
-      enableStream(RS_STREAM_COLOR, width_[RS_STREAM_COLOR], height_[RS_STREAM_COLOR], format_[RS_STREAM_COLOR],
-          fps_[RS_STREAM_COLOR]);
-      if (camera_info_ptr_[RS_STREAM_COLOR] == NULL)
+      publishTopic(RS_STREAM_DEPTH, frame);
+
+      if (enable_pointcloud_ == true)
       {
-        ROS_DEBUG_STREAM(nodelet_name_ << " - Allocating resources for " << STREAM_DESC[RS_STREAM_COLOR]);
-        getStreamCalibData(RS_STREAM_COLOR);
-        step_[RS_STREAM_COLOR] = camera_info_ptr_[RS_STREAM_COLOR]->width * unit_step_size_[RS_STREAM_COLOR];
-        image_[RS_STREAM_COLOR] = cv::Mat(camera_info_ptr_[RS_STREAM_COLOR]->height,
-            camera_info_ptr_[RS_STREAM_COLOR]->width, cv_type_[RS_STREAM_COLOR], cv::Scalar(0, 0, 0));
+        publishPCTopic();
       }
-      ts_[RS_STREAM_COLOR] = -1;
-    }
-    else if (enable_[RS_STREAM_COLOR] == false)
-    {
-      disableStream(RS_STREAM_COLOR);
-    }
+    };
 
-    if (enable_[RS_STREAM_DEPTH] == true)
+    color_frame_handler_ = [&](rs::frame  frame)
     {
-      enableStream(RS_STREAM_DEPTH, width_[RS_STREAM_DEPTH], height_[RS_STREAM_DEPTH], format_[RS_STREAM_DEPTH],
-          fps_[RS_STREAM_DEPTH]);
-      if (camera_info_ptr_[RS_STREAM_DEPTH] == NULL)
-      {
-        ROS_DEBUG_STREAM(nodelet_name_ << " - Allocating resources for " << STREAM_DESC[RS_STREAM_DEPTH]);
-        getStreamCalibData(RS_STREAM_DEPTH);
-        step_[RS_STREAM_DEPTH] = camera_info_ptr_[RS_STREAM_DEPTH]->width * unit_step_size_[RS_STREAM_DEPTH];
-        image_[RS_STREAM_DEPTH] = cv::Mat(camera_info_ptr_[RS_STREAM_DEPTH]->height,
-            camera_info_ptr_[RS_STREAM_DEPTH]->width, cv_type_[RS_STREAM_DEPTH], cv::Scalar(0, 0, 0));
-      }
-      ts_[RS_STREAM_DEPTH] = -1;
-      depth_scale_meters_ = rs_get_device_depth_scale(rs_device_, &rs_error_);
-      checkError();
-    }
-    else if (enable_[RS_STREAM_DEPTH] == false)
+      publishTopic(RS_STREAM_COLOR, frame);
+    };
+
+    ir_frame_handler_ = [&](rs::frame  frame)
     {
-      disableStream(RS_STREAM_DEPTH);
-    }
+      publishTopic(RS_STREAM_INFRARED, frame);
+    };
+
+    rs_set_frame_callback_cpp(rs_device_, RS_STREAM_DEPTH, new rs::frame_callback(depth_frame_handler_), &rs_error_);
+    checkError();
 
-    if (enable_[RS_STREAM_INFRARED] == true)
+    rs_set_frame_callback_cpp(rs_device_, RS_STREAM_COLOR, new rs::frame_callback(color_frame_handler_), &rs_error_);
+    checkError();
+
+    rs_set_frame_callback_cpp(rs_device_, RS_STREAM_INFRARED, new rs::frame_callback(ir_frame_handler_), &rs_error_);
+    checkError();
+  }
+
+  /*
+   * Set the streams according to their corresponding flag values.
+   */
+  void BaseNodelet::setStreams()
+  {
+    // Enable streams
+    for (int stream=0; stream < STREAM_COUNT; stream++)
```

Comment:
We cant change STREAM_COUNT to an rs_stream type, since we are setting the value to 5 in constants.h. I would like us to actually remove the constant in a future patch and do this a different way, as I agree this is non-ideal. I would prefer a member variable num_streams_ that isn't hard-coded and varies depending on the camera type, set in the onInit() function for each type to the correct value for that camera. But that's another patch for another day.